### PR TITLE
perf: run batch inversion concurrently per fragment

### LINKED ISSUE #2405: Run batch inversion concurrently, per fragment
Context: https://github.com/0xMiden/miden-vm/pull/2345#discussion_r2576778297

### PR DESCRIPTION
**Pull request by Farukest**
Description:
## Describe your changes

Run batch inversion on the H0 helper column concurrently using `par_chunks_mut` per fragment, instead of processing the entire column sequentially.

Closes #2405

## Checklist before requesting a review
- [x] Repo forked and branch created from `next` according to naming convention.
- [x] Commit messages and codestyle follow [conventions](../CONTRIBUTING.md).
- [x] Commits are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- [x] Relevant issues are linked in the PR description.
- [x] Tests added for new functionality.
- [x] Documentation/comments updated according to changes.
- [x] Updated `CHANGELOG.md`

### CODE REVIEW

**Comment by Farukest** [comment]
All done @plafer 
Need your review. Thanks :)

File: processor/src/parallel/mod.rs [hunk 1]
```diff
@@ -249,9 +249,13 @@ fn generate_core_trace_columns(
...
-    // Run batch inversion on stack's H0 helper column
-    core_trace_columns[STACK_TRACE_OFFSET + H0_COL_IDX] =
-        batch_inversion(&core_trace_columns[STACK_TRACE_OFFSET + H0_COL_IDX]);
+    // Run batch inversion on stack's H0 helper column concurrently, per fragment
+    core_trace_columns[STACK_TRACE_OFFSET + H0_COL_IDX]
+        .par_chunks_mut(fragment_size)
+        .for_each(|chunk| {
+            let inverted = batch_inversion(chunk);
+            chunk.copy_from_slice(&inverted);
+        });
```
Comment (plafer) [inline processor/src/parallel/mod.rs]: While this works, I would do this `batch_inversion()` in `CoreTraceFragmentFiller::fill_fragment()`, right after the call to `self.fill_fragment_impl(execution_state)`.

Basically

```rs
  pub fn fill_fragment(mut self) -> ([Felt; STACK_TRACE_WIDTH], [Felt; SYS_TRACE_WIDTH], usize) {
      let execution_state = self.context.initial_execution_state.clone();
      // Execute fragment generation and always finalize at the end
      let _ = self.fill_fragment_impl(execution_state);

      // Invert H0 column in the stack trace
      let num_rows_built = self.num_rows_built();
      {
          let inverted_h0 = batch_inversion(
              &self.fragment.columns[STACK_TRACE_OFFSET + H0_COL_IDX][0..self.num_rows_built()],
          );
          self.fragment.columns[STACK_TRACE_OFFSET + H0_COL_IDX][0..num_rows_built]
              .copy_from_slice(&inverted_h0);
      }

      let final_stack_rows = self.stack_rows.unwrap_or([ZERO; STACK_TRACE_WIDTH]);
      let final_system_rows = self.system_rows.unwrap_or([ZERO; SYS_TRACE_WIDTH]);
      (final_stack_rows, final_system_rows, num_rows_built)
  }
```

Comment (Farukest) [reply processor/src/parallel/mod.rs]: Its good.. Done @plafer moved `batch_inversion()` into `fill_fragment()` as suggested.

**Comment by Farukest** [comment]
Hey :)
any update @plafer ?

**Comment by plafer** [comment]
@Farukest can you check why the tests are failing now?

Since all but 2 tests are passing, I wonder if there's an edge case that we're inverting 1 extra/fewer element in certain conditions.

**Comment by Farukest** [comment]
> can you check why the tests are failing now?
> Since all but 2 tests are passing, I wonder if there's an edge case that we're inverting 1 extra/fewer element in certain conditions.

I see. I'm checking now @plafer

**Comment by Farukest** [comment]
> While this works, I would do this batch_inversion() in CoreTraceFragmentFiller::fill_fragment(), right after the call to self.fill_fragment_impl(execution_state).

@plafer I tried to apply your suggestion but moving `batch_inversion()` into `fill_fragment()` causes test failures.

The issue is the execution order:
1. `fill_fragment()` runs and inverts H0 for rows in this fragment
2. `fixup_stack_and_system_rows()` then overwrites the **first row** of each fragment with the previous fragment's buffer
3. This buffer contains **non-inverted** H0 values (from `stack_rows` returned by `fill_fragment`)

So row _65536_ (first row of fragment 2) gets overwritten with a **non inverted** value after being **inverted**.

The fix must run `batch_inversion` **after** `fixup_stack_and_system_rows()` completes. We still get parallelism via `par_chunks_mut()`.

The failing tests showed:
- `stack_h0` at row 65536: slow=17831852600434098177 ( inverted ), parallel=30 ( non inverted )

**Comment by plafer** [comment]
Ah yes that makes sense - thank you for the analysis!

In that case let's revert to what you had initially.

**Comment by Farukest** [comment]
> Ah yes that makes sense - thank you for the analysis!
> In that case let's revert to what you had initially.

Already pushed @plafer . Done ! :)

**Comment by plafer** [comment]
> Already pushed @plafer . Done ! :)

Did you push the intended commits? I still see the incorrect implementation, and tests are still failing

**Comment by Farukest** [comment]
> Did you push the intended commits? I still see the incorrect implementation, and tests are still failing

My mistake. Fixed now.

**Comment by bobbinth** [comment]
@Farukest - thank you for tackling this! I'm curious, did this have any noticeable impact on performance?

**Comment by plafer** [comment]
I checked, and unfortunately it didn't. I figured we might as well merge it, if not only so that we don't have to worry about that being a performance penalty.
